File: 2025/1/week1/휴대폰 자판_백준_5670/Solution.py
class Node:
    def __init__(self, character, endOfWord = False):
        self.endOfWord = endOfWord
        self.child = {}
        self.chracter = character

class Trie:
    def __init__(self):
        self.root = Node("")
    
    def insert(self, word):
        curNode = self.root
        for chracter in word:
            if chracter not in curNode.child:
                newNode = Node(chracter)
                curNode.child[chracter]=newNode
            curNode = curNode.child[chracter]
        curNode.endOfWord = True
    
    def find(self, word):
        curNode = self.root
        for chracter in word:
            if chracter not in curNode.child:
                return False
            curNode = curNode.child[chracter]
        if not curNode.endOfWord:
            return False
        
        return True
    
    def printTrie(self, node:Node, depth):
        print(depth*"-"+node.chracter+"|"+str(node.endOfWord))
        if len(node.child) > 0:
            for _, childNode in node.child.items():
                self.printTrie(childNode,depth+1)

    def calcClickCount(self, node:Node, click):
        totalCount = 0
        if len(node.child) > 0:
            for _, childNode in node.child.items():
                if len(node.child) > 1 or node.endOfWord:
                    totalCount+= self.calcClickCount(childNode,click+1)
                else:
                    totalCount+= self.calcClickCount(childNode,click)
        if node.endOfWord:
            # print(click)
            totalCount+=click
        return totalCount
    
import sys

input = sys.stdin.readline

while True:
    try:
        n = int(input())
    except Exception as e:
        break
    trie = Trie()
    totalCount = 0
    for _ in range(n):
        trie.insert(input().strip())
    for _, childNode in trie.root.child.items():
        totalCount+= trie.calcClickCount(childNode, 1)
    # trie.printTrie(trie.root,0)
    # print(totalCount)
    print(f"{totalCount/n:.2f}")

File: 2025/1/week1/휴대폰 자판_백준_5670/Solution.java
import java.io.*;
import java.util.*;

class Solution{
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    public static void main(String[] args) throws IOException{
        while(true){
            int n = 0;
            try{
                String input = br.readLine();
                n = Integer.parseInt(input);
            }catch(Exception e){
                break;
            }
            Trie trie = new Trie();
            for (int i = 0; i < n; i++) {
                trie.insert(br.readLine().trim());
            }
            System.out.printf("%.02f\n",trie.getClickAverage(n));
        }
    }
}

class Node{
    Character charcter;
    HashMap<Character, Node> child;
    boolean endOfWord;

    Node(char charcter){
        child = new HashMap<>();
        endOfWord = false;
        this.charcter = charcter;
    }
}
class Trie{
    Node root;

    public Trie(){
        root = new Node(' ');
    }

    public void insert(String word){
        Node curNode = root;
        char[] charOfWord = word.toCharArray();
        for (char curChar : charOfWord) {
            if (!curNode.child.containsKey(curChar)) {
                curNode.child.put(curChar, new Node(curChar));
            }
            curNode = curNode.child.get(curChar);
        }
        curNode.endOfWord = true;
    }

    public float getClickAverage(int n){
        Collection<Node> childs = root.child.values();
        int total = 0;
        for (Node node : childs) {
            total+=calcClickCount(node, 1);
        }
        return (float) total /n;
    }

    private int calcClickCount(Node node, int click){
        int totalClickCount = 0;
        Collection<Node> curChilds = node.child.values();
        for (Node childNode : curChilds) {
            if(node.child.size() > 1 || node.endOfWord){
                totalClickCount += calcClickCount(childNode, click+1);
            }else{
                totalClickCount += calcClickCount(childNode, click);
            }
        }

        if(node.endOfWord){
            totalClickCount += click;
        }
        return totalClickCount;
    }
}